### ISSUE TITLE
Random occasional `SubscriptionWithCancelException$NoMoreElementsNeeded` after migrating from Akka to Pekko

### ISSUE THREAD

**Issue by Tvaroh** [apache/incubator-pekko-http#422]
Hey, I notice that when I migrate to Pekko (form latest "free" Akka), I start getting occasional `org.apache.pekko.stream.SubscriptionWithCancelException$NoMoreElementsNeeded$ (No error message supplied)` from web clients.

I've increases `stream-cancellation-delay` to `1s` but still it fails sometimes. It does really look like something changed from Akka to Pekko.

I don't really have any code to reproduce as it's not consistent, maybe once in couple of days. But reverting back to Akka does fix it.

**Comment by He-Pin** [comment]
Do you mean it works after you revert back to akka 2.6.20？

**Comment by Tvaroh** [comment]
@He-Pin yep.

**Comment by He-Pin** [comment]
Can you share a simple reproducer, thanks @Tvaroh

**Comment by Tvaroh** [comment]
As I said, it's inconsistent and happens sporadically. So I don't have a reproducer, but in the end it calls `Http().singleRequest(request)` which fails.

**Comment by jrudolph** [comment]
I don't know about any change which could have caused this issue. Can you still validate that the old version based on Akka is still fine (to rule out a coincidence where something else in the wider environment might have changed in the meantime)?

Your best bet is trying to collect debug log message from `org.apache.pekko.http.impl.engine.client.pool.NewHostConnectionPool` to isolate the issue.

Since the issue indeed looks like it could be caused by the infamous stream cancellation race condition, you could try setting `stream-cancellation-delay` to much higher values to rule out any race-conditions that might still occur in a 1 second time frame caused by GC or other interruptions.

What kind of requests run into that problem? Usually, idempotent requests are retried a few times, which would make it even more unlikely to happen. So, this might also be an issue of unfortunate timing, where an non-idempotent request (e.g. POST) is run at exactly the same time as the server decides to close the connection (however, that scenario is reported in an unclear way caused by stream cancellation). In that case, playing around with `idle-timeout` and `host-connection-pool.keep-alive-timeout` (see also its documentation in the `reference.conf`) might help mitigate the issue.

**Comment by Tvaroh** [comment]
@jrudolph yeah it only happens for requests to a particular internal service written in (God forgive) C#, so this can be some weird corner case, but it def don't happen on Akka. I'll try your suggestions and will keep an eye on it. Thanks for the reply, I'll apply some config changes and will post back.

**Comment by Tvaroh** [comment]
@djjorjinho it's mysterious but after adding retries those exceptions eventually went away by themselves, perhaps something changed externally, so I'm afraid I have no answer to that.

**Comment by djjorjinho** [comment]
@Tvaroh no worries, thanks for the reply.

Can you share if you made any changes to any config inside `pekko.http.host-connection-pool` or `pekko.http.client`, like `stream-cancellation-delay`, `idle-timeout`, `keep-alive-timeout`, `max-connection-lifetime`, etc.?

**Comment by Tvaroh** [comment]
@djjorjinho good point, sure, here what we ended up with:
```hocon
pekko {
  loggers = ["org.apache.pekko.event.slf4j.Slf4jLogger"]
  loglevel = "DEBUG"

  http {
    client {
      stream-cancellation-delay = 2s

      parsing {
        max-content-length = 128m
      }
    }

    server {
      stream-cancellation-delay = 2s
      request-timeout = 60s
    }

    parsing {
      max-uri-length = 8k
      max-header-value-length = 8k
      max-to-strict-bytes = 64m
    }

    routing {
      decode-max-size = 16m
    }

    host-connection-pool {
      max-open-requests = 64
      response-entity-subscription-timeout = 10.seconds
    }
  }
}

```

**Comment by djjorjinho** [comment]
In the end, I enabled debug logging, but still could not fully understand what was happening, except that the server our service was requesting had already responded with an entity payload and it looked like the issue was at consuming the payload.

I found this [discussion](https://discuss.lightbend.com/t/about-nomoreelementsneeded-exception/8599/6) and I noticed that we had increased `max-chunk-size-20mb` to this k8s api cluster discovery [issue](https://doc.akka.io/docs/akka-management/current/bootstrap/troubleshooting.html#kubernetes-api-http-chunk-size-exceeds-the-configured-limit-during-contact-point-discovery).

We use seed nodes for discovery now, so once we reverted the value back to 1mb, the exceptions stopped.

**Comment by JD557** [comment]
Just to add an extra note - this was a problem in Akka as well: https://github.com/akka/akka-http/issues/3201

**Comment by mdedetrich** [comment]
So I can still finish off the PR in Pekko as it appears that I basically didn't finish off the PR (changing the value caused some tests to fail and I needed to do adjustments to fix those tests)

**Comment by mdedetrich** [comment]
PR made  at https://github.com/apache/pekko-http/pull/590, still need to fix the tests still they don't expect such a long stream-cancellation-delay

**Comment by mdedetrich** [comment]
With Pekko being forked off of Akka 2.6.x it seems that we can now solve this properly i.e.

> IIRC, the whole reason to introduce the delay was to deal with the fact that stream cancellation could not be well-configured in Akka 2.5. With 2.6, it's not 100% clear if the delay is even still needed or whether it could be solved in a better way (i.e. by resolving the cancellation race more intelligently, taking the cancellation reason into account at the right/more places) to avoid symptoms as shown in the ticket.

_Originally posted by @jrudolph in https://github.com/apache/pekko-http/pull/590#discussion_r1747075727_

**Comment by pjfanning** [comment]
#590 has been reverted (#597). We can revisit this for v1.1.1.
